waku/v2/waku_core/topics/sharding.nim

```diff
@@ -0,0 +1,113 @@
+## Waku autosharding utils
+##
+## See 51/WAKU2-RELAY-SHARDING RFC: https://rfc.vac.dev/spec/51/#automatic-sharding
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  nimcrypto,
+  std/math,
+  std/sequtils,
+  std/algorithm,
+  std/strutils,
+  stew/endians2,
+  stew/results,
+  stew/byteutils
+
+import
+  ./content_topic,
+  ./pubsub_topic
+
+const ClusterIndex* = 49152
+const GenerationZeroShardsCount* = 5
+
+type ShardsPriority = seq[tuple[topic: NsPubsubTopic, value: float64]]
+
+proc applyWeight(hashValue: uint64, weight: float64): float64 =
+  -weight / math.ln(float64(hashValue) / float64(high(uint64)))
+
+proc hashOrder*(x, y: (NsPubsubTopic, float64)): int =
+    cmp(x[1], y[1])
+
+proc weightedShardList*(topic: NsContentTopic, shardCount: int, weights: seq[float64]): Result[ShardsPriority, string] =
+  ## Returns the ordered list of shards and their priority values.
+
+  if weights.len != shardCount:
+    return err("Must provide weights for every shards")
+
+  let shardsNWeights = zip(toSeq(0..shardCount), weights)
+
+  var list = newSeq[(NsPubsubTopic, float64)](shardCount)
+
+  for (shard, weight) in shardsNWeights:
+    let pubsub = NsPubsubTopic.staticSharding(ClusterIndex, uint16(shard))
+    let bytes = toBytes($topic) & toBytes($pubsub)
+    let hash = sha256.digest(bytes)
+    let hashValue = uint64.fromBytesBE(hash.data)
+    let value = applyWeight(hashValue, weight)
+
+    list[shard] = (pubsub, value)
+
+  list.sort(hashOrder)
+
+  ok(list)
+
+type ShardingBias* = enum
+  None = "none"
+  Kanonymity = "anon"
+  Throughput = "bandwidth"
```

Comment:
Yes that's the assumption.

Since we can't control the "population" of the shards with or without biasing the shard selection we may end with "pathological" cases.